crates/sui-json-rpc/src/read_api.rs

```diff
@@ -1240,21 +1259,70 @@ async fn get_display_fields(
     original_object: &Object,
     original_layout: &Option<MoveStructLayout>,
 ) -> Result<DisplayFieldsResponse, ObjectDisplayError> {
-    let Some((object_type, layout)) = get_object_type_and_struct(original_object, original_layout)?
+    let Some(layout) = original_layout else {
+        return Ok(DisplayFieldsResponse {
+            data: None,
+            error: None,
+        });
+    };
+
+    let Some(move_object) = original_object.data.try_as_move() else {
+        return Err(ObjectDisplayError::MoveObject);
+    };
+
+    let Some(display_object) =
+        get_display_object_by_type(kv_store, fullnode_api, &layout.type_).await?
     else {
         return Ok(DisplayFieldsResponse {
             data: None,
             error: None,
         });
     };
-    if let Some(display_object) =
-        get_display_object_by_type(kv_store, fullnode_api, &object_type).await?
-    {
-        return get_rendered_fields(display_object.fields, &layout);
+
+    let format = match Format::parse(MAX_DISPLAY_NESTED_LEVEL, &display_object.fields) {
+        Ok(format) => format,
+        Err(e) => {
+            return Ok(DisplayFieldsResponse {
+                data: None,
+                error: Some(SuiObjectResponseError::DisplayError {
+                    error: e.to_string(),
+                }),
+            });
+        }
+    };
+
+    let layout = MoveTypeLayout::Struct(Box::new(layout.clone()));
+    let display = match format.display(*MAX_DISPLAY_OUTPUT_SIZE, move_object.contents(), &layout) {
+        Ok(fields) => fields,
+        Err(e) => {
+            return Ok(DisplayFieldsResponse {
+                data: None,
+                error: Some(SuiObjectResponseError::DisplayError {
+                    error: e.to_string(),
+                }),
+            });
+        }
+    };
+
+    let mut fields = BTreeMap::new();
```

Comment:
The BTreeMap is sort of overkill given json key order in objects is explicitly unordered per the spec

Reply:
This is pre-existing though, right? The choice of data structure comes from the existing `DisplayFieldsResponse` type. 

Having said that, I do tend to default to `BTreeMap` because having a stable output ordering is helpful for snapshot testing.

Reply:
no the comment is more about relying on any client to see the same order, per spec that's not guaranteed. I missed that this was pre-existing though

Reply:
Ah right, yeah there is no guarantee to clients on order (we should be able to change the order, and I think I have done in the past).